File: frontend/src/types/camera/camera.ts
import type { CameraCapabilities } from './capabilities';
import type { CameraProfile } from './stream';

// Base camera model synchronized with backend
export interface Camera {
  id: string;
  name: string;
  url: string;
  username: string;
  password: string;
  protocol: CameraProtocol;
  status: CameraStatus;
  capabilities?: CameraCapabilities;
  profiles: CameraProfile[];
  createdAt: string;
  lastConnectedAt: string;
  deviceInfo?: CameraDeviceInfo;
}

export interface CameraDeviceInfo {
  manufacturer?: string;
  model?: string;
  firmwareVersion?: string;
  serialNumber?: string;
}

export interface CameraCredentials {
  username: string;
  password: string;
}

export enum CameraStatus {
  Offline = 0,      // Camera is not connected
  Connecting = 1,   // Camera is in the process of connecting
  Online = 2,       // Camera is connected and healthy
  Degraded = 3,     // Camera is connected but experiencing health issues
  Error = 4         // Camera has encountered an error
}

export enum CameraProtocol {
  Onvif = 0,
  Rtsp = 1
}

// Request models
export interface AddCameraRequest {
  name: string;
  url: string;
  username: string;
  password: string;
  protocol: CameraProtocol;
}

export interface UpdateCameraRequest {
  id: string;
  name: string;
  url: string;
  username: string;
  password: string;
  protocol: CameraProtocol;
}

// Response models
export interface CameraStatusResponse {
  isOnline: boolean;
  status: string;
  lastSeen?: string;
  errorMessage?: string;
}

File: shared/SharedContracts/Camera.cs
namespace Lightview.Shared.Contracts;

// Camera Management Models
public class Camera
{
    public Guid Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public required Uri Url { get; set; }
    public string Username { get; set; } = string.Empty;
    public string Password { get; set; } = string.Empty;
    public CameraProtocol Protocol { get; set; } = CameraProtocol.Onvif;
    public CameraStatus Status { get; set; } = CameraStatus.Offline;
    public CameraCapabilities? Capabilities { get; set; }
    public List<CameraProfile> Profiles { get; set; } = new();
    public DateTime CreatedAt { get; set; }
    public DateTime LastConnectedAt { get; set; }
    public CameraDeviceInfo? DeviceInfo { get; set; }
}

public class CameraDeviceInfo
{
    public string? Manufacturer { get; set; }
    public string? Model { get; set; }
    public string? FirmwareVersion { get; set; }
    public string? SerialNumber { get; set; }
}

public class CameraCredentials
{
    public string Username { get; set; } = string.Empty;
    public string Password { get; set; } = string.Empty;
}

public enum CameraStatus
{
    Offline,     // Camera added but no connection attempts made
    Connecting,  // Camera-controller attempting to connect
    Online,      // Camera connected successfully, no health check failures
    Degraded,    // Camera connected but health checks failing
    Error        // Connection failed or unrecoverable error
}

public enum CameraProtocol
{
    Onvif,
    Rtsp
}

// Health Check Models
public class HealthCheckResult
{
    public bool IsSuccessful { get; set; }
    public string CheckName { get; set; } = string.Empty;
    public string? ErrorMessage { get; set; }
    public TimeSpan ResponseTime { get; set; }
    public DateTime CheckedAt { get; set; } = DateTime.UtcNow;
}

public class CameraHealthCheckResults
{
    public bool OverallHealthy { get; set; }
    public List<HealthCheckResult> Results { get; set; } = new();
    public DateTime CheckedAt { get; set; } = DateTime.UtcNow;
    public TimeSpan TotalResponseTime { get; set; }
}
